In [1]:
import pandas as pd

<h3>Collecting and loading the dataset</h3>

*  I have collected two datasets containing details about both phishing and legitimate websites.
*  Each dataset includes features of the URLs along with corresponding labels:
     *  A label of 0 indicates a legitimate website.
     *  A label of 1 indicates a phishing website.
* The data is based on HTML content and includes various features extracted from the URLs.
 URLs

In [2]:
legit = pd.read_csv("structured_data_legitimate.csv")

In [3]:
legit.head()

has_title  has_input  has_button  has_image  has_submit  has_link  \
0          1          1           0          0           1         0   
1          1          1           0          0           0         1   
2          1          1           1          0           0         1   
3          1          1           1          0           0         1   
4          1          1           1          0           0         1   

   has_password  has_email_input  has_hidden_element  has_audio  ...  \
0             0                0                   1          0  ...   
1             0                0                   0          0  ...   
2             1                0                   1          0  ...   
3             0                0                   0          0  ...   
4             0                1                   0          0  ...   

   has_text_input  number_of_meta  has_nav  has_object  has_picture  \
0               0               2        0           0            0   
1               1               7        0           0            0   
2               1              18        0           0            0   
3               0              26        1           0            1   
4               0              20        0           0            0   

   number_of_sources  number_of_span  number_of_table                   URL  \
0                  0               8                1     http://google.com   
1                  0               1                0    http://youtube.com   
2                  0               2                0   http://facebook.com   
3                 44              27                0  http://microsoft.com   
4                  2              31                0    http://netflix.com   

   label  
0      0  
1      0  
2      0  
3      0  
4      0  

[5 rows x 45 columns]

In [4]:
phish = pd.read_csv("structured_data_phishing.csv")

In [5]:
phish.head()

has_title  has_input  has_button  has_image  has_submit  has_link  \
0          0          0           0          0           0         0   
1          0          0           0          0           0         0   
2          0          0           0          0           0         0   
3          0          0           0          0           0         0   
4          0          0           0          0           0         0   

   has_password  has_email_input  has_hidden_element  has_audio  ...  \
0             0                0                   0          0  ...   
1             0                0                   0          0  ...   
2             0                0                   0          0  ...   
3             0                0                   0          0  ...   
4             0                0                   0          0  ...   

   has_text_input  number_of_meta  has_nav  has_object  has_picture  \
0               0               0        0           0            0   
1               0               0        0           0            0   
2               0               0        0           0            0   
3               0               0        0           0            0   
4               0               0        0           0            0   

   number_of_sources  number_of_span  number_of_table  \
0                  0               0                0   
1                  0               0                0   
2                  0               0                0   
3                  0               0                0   
4                  0               0                0   

                                                 URL  label  
0  http://www.eki-net.con-aescceccesaas.qfuhtb.to...      1  
1  http://www.eki-net.con-aescceccesaas.qsnbpy.to...      1  
2  http://www.eki-net.con-aescceccesaas.trea.top/...      1  
3  http://www.eki-net.con-aescceccesaas.ucwxyw.to...      1  
4  http://www.eki-net.con-aescceccesaas.vcuu.top/...      1  

[5 rows x 45 columns]

In [6]:
phish.shape

(10524, 45)

In [7]:
legit.shape

(16061, 45)

In [8]:
legit.columns

Index(['has_title', 'has_input', 'has_button', 'has_image', 'has_submit',
       'has_link', 'has_password', 'has_email_input', 'has_hidden_element',
       'has_audio', 'has_video', 'number_of_inputs', 'number_of_buttons',
       'number_of_images', 'number_of_option', 'number_of_list',
       'number_of_th', 'number_of_tr', 'number_of_href', 'number_of_paragraph',
       'number_of_script', 'length_of_title', 'has_h1', 'has_h2', 'has_h3',
       'length_of_text', 'number_of_clickable_button', 'number_of_a',
       'number_of_img', 'number_of_div', 'number_of_figure', 'has_footer',
       'has_form', 'has_text_area', 'has_iframe', 'has_text_input',
       'number_of_meta', 'has_nav', 'has_object', 'has_picture',
       'number_of_sources', 'number_of_span', 'number_of_table', 'URL',
       'label'],
      dtype='object')

<h4>Changing the order of the URL column</h4>

*   I am rearranging the order of the URL column to facilitate the addition of new features to the end
d

In [9]:
url_column = legit['URL']

In [10]:
legit.drop(columns=["URL"],inplace=True)

In [11]:
legit.insert(0,"URL",url_column)


In [12]:
legit

URL  has_title  has_input  has_button  has_image  \
0            http://google.com          1          1           0          0   
1           http://youtube.com          1          1           0          0   
2          http://facebook.com          1          1           1          0   
3         http://microsoft.com          1          1           1          0   
4           http://netflix.com          1          1           1          0   
...                        ...        ...        ...         ...        ...   
16056       http://emis.gov.eg          1          1           0          0   
16057     http://hilton.com.tr          1          1           1          0   
16058  http://medicines.org.uk          1          1           1          0   
16059        http://piwigo.org          1          0           1          0   
16060        http://koddos.com          1          0           0          0   

       has_submit  has_link  has_password  has_email_input  \
0               1         0             0                0   
1               0         1             0                0   
2               0         1             1                0   
3               0         1             0                0   
4               0         1             0                1   
...           ...       ...           ...              ...   
16056           0         1             0                0   
16057           0         1             0                0   
16058           0         1             0                0   
16059           0         1             0                0   
16060           0         0             0                0   

       has_hidden_element  ...  has_iframe  has_text_input  number_of_meta  \
0                       1  ...           0               0               2   
1                       0  ...           1               1               7   
2                       1  ...           0               1              18   
3                       0  ...           1               0              26   
4                       0  ...           0               0              20   
...                   ...  ...         ...             ...             ...   
16056                   1  ...           0               0               1   
16057                   1  ...           0               0               5   
16058                   0  ...           1               1               6   
16059                   0  ...           0               0               4   
16060                   0  ...           0               0               0   

       has_nav  has_object  has_picture  number_of_sources  number_of_span  \
0            0           0            0                  0               8   
1            0           0            0                  0               1   
2            0           0            0                  0               2   
3            1           0            1                 44              27   
4            0           0            0                  2              31   
...        ...         ...          ...                ...             ...   
16056        1           0            0                  0               4   
16057        1           0            1                 22               9   
16058        1           0            0                  0              27   
16059        1           0            0                  0              21   
16060        0           0            0                  0               0   

       number_of_table  label  
0                    1      0  
1                    0      0  
2                    0      0  
3                    0      0  
4                    0      0  
...                ...    ...  
16056                2      0  
16057                0      0  
16058                0      0  
16059                0      0  
16060                0      0  

[16061 rows x 45 columns]

In [13]:
phish_url = phish['URL']

In [14]:
phish.drop(columns=['URL'],inplace=True)

In [15]:
phish.insert(0,'URL',phish_url)

In [16]:
phish

URL  has_title  \
0      http://www.eki-net.con-aescceccesaas.qfuhtb.to...          0   
1      http://www.eki-net.con-aescceccesaas.qsnbpy.to...          0   
2      http://www.eki-net.con-aescceccesaas.trea.top/...          0   
3      http://www.eki-net.con-aescceccesaas.ucwxyw.to...          0   
4      http://www.eki-net.con-aescceccesaas.vcuu.top/...          0   
...                                                  ...        ...   
10519                              https://dloodles.app/          1   
10520                             https://artblooks.xyz/          1   
10521                           http://ohnizwl.htswyx.cn          1   
10522                https://poocoinect-tk.blogspot.com/          1   
10523                    https://osmosis.sezenradyo.com/          1   

       has_input  has_button  has_image  has_submit  has_link  has_password  \
0              0           0          0           0         0             0   
1              0           0          0           0         0             0   
2              0           0          0           0         0             0   
3              0           0          0           0         0             0   
4              0           0          0           0         0             0   
...          ...         ...        ...         ...       ...           ...   
10519          0           0          0           0         0             0   
10520          0           0          0           0         1             0   
10521          0           0          0           0         0             0   
10522          0           1          0           0         1             0   
10523          0           1          0           0         1             0   

       has_email_input  has_hidden_element  ...  has_iframe  has_text_input  \
0                    0                   0  ...           0               0   
1                    0                   0  ...           0               0   
2                    0                   0  ...           0               0   
3                    0                   0  ...           0               0   
4                    0                   0  ...           0               0   
...                ...                 ...  ...         ...             ...   
10519                0                   0  ...           0               0   
10520                0                   0  ...           0               0   
10521                0                   0  ...           0               0   
10522                0                   0  ...           0               0   
10523                0                   0  ...           0               0   

       number_of_meta  has_nav  has_object  has_picture  number_of_sources  \
0                   0        0           0            0                  0   
1                   0        0           0            0                  0   
2                   0        0           0            0                  0   
3                   0        0           0            0                  0   
4                   0        0           0            0                  0   
...               ...      ...         ...          ...                ...   
10519               0        0           0            0                  0   
10520              11        0           0            0                  0   
10521               1        0           0            0                  0   
10522               4        1           0            0                  0   
10523               3        1           0            0                  0   

       number_of_span  number_of_table  label  
0                   0                0      1  
1                   0                0      1  
2                   0                0      1  
3                   0                0      1  
4                   0                0      1  
...               ...              ...    ...  
10519               0        

In [17]:
phish_url.head()

0    http://www.eki-net.con-aescceccesaas.qfuhtb.to...
1    http://www.eki-net.con-aescceccesaas.qsnbpy.to...
2    http://www.eki-net.con-aescceccesaas.trea.top/...
3    http://www.eki-net.con-aescceccesaas.ucwxyw.to...
4    http://www.eki-net.con-aescceccesaas.vcuu.top/...
Name: URL, dtype: object

In [18]:
phish_url.tail(10)

10514                                  http://Shido.market
10515                                https://tmac.finance/
10516                                 https://web-hft.com/
10517                             https://tiny.one/pznd63c
10518    https://synslabs.firebaseapp.com/wallet/connec...
10519                                https://dloodles.app/
10520                               https://artblooks.xyz/
10521                             http://ohnizwl.htswyx.cn
10522                  https://poocoinect-tk.blogspot.com/
10523                      https://osmosis.sezenradyo.com/
Name: URL, dtype: object

In [19]:
phish.drop("label",inplace=True,axis=1)

In [20]:
legit.drop("label",inplace=True,axis=1)

<h4>Importing packages needed to extract features from the url's</h4>

In [21]:
from urllib.parse import urlparse,urlencode
import ipaddress


In [22]:

from requests.packages.urllib3.util.retry import Retry
from urllib3.exceptions import InsecureRequestWarning
from urllib3 import disable_warnings
import urllib.request

<h3>Length of URL</h3>

*  Computes the length of the URL. Phishers can use long URL to hide the doubtful part in the address bar. In this project, if the length of the URL is    greater than or equal 54 characters then the URL classified as phishing otherwise legitimate.
*  If the length of URL >= 54 , the value assigned to this feature is 1 (phishing) or else 0 (legitimate).

In [23]:
def get_url_length(url):
    if len(url) >= 54:
        return 1
    else:
        return 0

In [24]:
phish["url_length"] = phish["URL"].apply(get_url_length)

In [25]:
phish

URL  has_title  \
0      http://www.eki-net.con-aescceccesaas.qfuhtb.to...          0   
1      http://www.eki-net.con-aescceccesaas.qsnbpy.to...          0   
2      http://www.eki-net.con-aescceccesaas.trea.top/...          0   
3      http://www.eki-net.con-aescceccesaas.ucwxyw.to...          0   
4      http://www.eki-net.con-aescceccesaas.vcuu.top/...          0   
...                                                  ...        ...   
10519                              https://dloodles.app/          1   
10520                             https://artblooks.xyz/          1   
10521                           http://ohnizwl.htswyx.cn          1   
10522                https://poocoinect-tk.blogspot.com/          1   
10523                    https://osmosis.sezenradyo.com/          1   

       has_input  has_button  has_image  has_submit  has_link  has_password  \
0              0           0          0           0         0             0   
1              0           0          0           0         0             0   
2              0           0          0           0         0             0   
3              0           0          0           0         0             0   
4              0           0          0           0         0             0   
...          ...         ...        ...         ...       ...           ...   
10519          0           0          0           0         0             0   
10520          0           0          0           0         1             0   
10521          0           0          0           0         0             0   
10522          0           1          0           0         1             0   
10523          0           1          0           0         1             0   

       has_email_input  has_hidden_element  ...  has_iframe  has_text_input  \
0                    0                   0  ...           0               0   
1                    0                   0  ...           0               0   
2                    0                   0  ...           0               0   
3                    0                   0  ...           0               0   
4                    0                   0  ...           0               0   
...                ...                 ...  ...         ...             ...   
10519                0                   0  ...           0               0   
10520                0                   0  ...           0               0   
10521                0                   0  ...           0               0   
10522                0                   0  ...           0               0   
10523                0                   0  ...           0               0   

       number_of_meta  has_nav  has_object  has_picture  number_of_sources  \
0                   0        0           0            0                  0   
1                   0        0           0            0                  0   
2                   0        0           0            0                  0   
3                   0        0           0            0                  0   
4                   0        0           0            0                  0   
...               ...      ...         ...          ...                ...   
10519               0        0           0            0                  0   
10520              11        0           0            0                  0   
10521               1        0           0            0                  0   
10522               4        1           0            0                  0   
10523               3        1           0            0                  0   

       number_of_span  number_of_table  url_length  
0                   0                0           1  
1                   0                0           1  
2                   0                0           0  
3                   0                0           1  
4                   0                0           0  
...               ...              ...         

In [49]:
phish["url_length"].value_counts()

url_length
1    6414
0    4110
Name: count, dtype: int64

In [50]:
legit["url_length"] = legit["URL"].apply(get_url_length)

In [51]:
legit

URL  has_title  has_input  has_button  has_image  \
0            http://google.com          1          1           0          0   
1           http://youtube.com          1          1           0          0   
2          http://facebook.com          1          1           1          0   
3         http://microsoft.com          1          1           1          0   
4           http://netflix.com          1          1           1          0   
...                        ...        ...        ...         ...        ...   
16056       http://emis.gov.eg          1          1           0          0   
16057     http://hilton.com.tr          1          1           1          0   
16058  http://medicines.org.uk          1          1           1          0   
16059        http://piwigo.org          1          0           1          0   
16060        http://koddos.com          1          0           0          0   

       has_submit  has_link  has_password  has_email_input  \
0               1         0             0                0   
1               0         1             0                0   
2               0         1             1                0   
3               0         1             0                0   
4               0         1             0                1   
...           ...       ...           ...              ...   
16056           0         1             0                0   
16057           0         1             0                0   
16058           0         1             0                0   
16059           0         1             0                0   
16060           0         0             0                0   

       has_hidden_element  ...  has_iframe  has_text_input  number_of_meta  \
0                       1  ...           0               0               2   
1                       0  ...           1               1               7   
2                       1  ...           0               1              18   
3                       0  ...           1               0              26   
4                       0  ...           0               0              20   
...                   ...  ...         ...             ...             ...   
16056                   1  ...           0               0               1   
16057                   1  ...           0               0               5   
16058                   0  ...           1               1               6   
16059                   0  ...           0               0               4   
16060                   0  ...           0               0               0   

       has_nav  has_object  has_picture  number_of_sources  number_of_span  \
0            0           0            0                  0               8   
1            0           0            0                  0               1   
2            0           0            0                  0               2   
3            1           0            1                 44              27   
4            0           0            0                  2              31   
...        ...         ...          ...                ...             ...   
16056        1           0            0                  0               4   
16057        1           0            1                 22               9   
16058        1           0            0                  0              27   
16059        1           0            0                  0              21   
16060        0           0            0                  0               0   

       number_of_table  url_length  
0                    1           0  
1                    0           0  
2                    0           0  
3                    0           0  
4                    0           0  
...                ...         ...  
16056                2           0  
16057                0           0  
16058                0           0  
16059                0           0  
16060                0           0  

[16061 rows x 

In [52]:
legit['url_length'].value_counts()

url_length
0    16060
1        1
Name: count, dtype: int64

 <h5>Data story</h5>

*  In the legitimate URLs, the majority (16060 instances) have a URL length less than a certain threshold, while only 1 instance has a URL               length greater than the threshold.
*  In the legitimate URLs, the majority (6414 instances) have a URL length greater than a certain threshold, while only (4110 instance) has a URL               length less than the thresholdd.

<h3>Prefix or Suffix "-" in Domain</h3>

*  Checking the presence of '-' in the domain part of URL. The dash symbol is rarely used in legitimate URLs. Phishers tend to add prefixes or suffixes    separated by (-) to the domain name so that users feel that they are dealing with a legitimate webpage.
*  If the URL has '-' symbol in the domain part of the URL, the value assigned to this feature is 1 (phishing) or else 0 (legitimate).

In [53]:
def has_hyphens(url):
    if '-' in urlparse(url).netloc: # we will be using the urlparse from urllib.parse for checking the presence of '-' in the url
        return 1
    else:
        return 0

In [54]:
legit["has_hyphens"] = legit["URL"].apply(has_hyphens)

In [55]:
legit

URL  has_title  has_input  has_button  has_image  \
0            http://google.com          1          1           0          0   
1           http://youtube.com          1          1           0          0   
2          http://facebook.com          1          1           1          0   
3         http://microsoft.com          1          1           1          0   
4           http://netflix.com          1          1           1          0   
...                        ...        ...        ...         ...        ...   
16056       http://emis.gov.eg          1          1           0          0   
16057     http://hilton.com.tr          1          1           1          0   
16058  http://medicines.org.uk          1          1           1          0   
16059        http://piwigo.org          1          0           1          0   
16060        http://koddos.com          1          0           0          0   

       has_submit  has_link  has_password  has_email_input  \
0               1         0             0                0   
1               0         1             0                0   
2               0         1             1                0   
3               0         1             0                0   
4               0         1             0                1   
...           ...       ...           ...              ...   
16056           0         1             0                0   
16057           0         1             0                0   
16058           0         1             0                0   
16059           0         1             0                0   
16060           0         0             0                0   

       has_hidden_element  ...  has_text_input  number_of_meta  has_nav  \
0                       1  ...               0               2        0   
1                       0  ...               1               7        0   
2                       1  ...               1              18        0   
3                       0  ...               0              26        1   
4                       0  ...               0              20        0   
...                   ...  ...             ...             ...      ...   
16056                   1  ...               0               1        1   
16057                   1  ...               0               5        1   
16058                   0  ...               1               6        1   
16059                   0  ...               0               4        1   
16060                   0  ...               0               0        0   

       has_object  has_picture  number_of_sources  number_of_span  \
0               0            0                  0               8   
1               0            0                  0               1   
2               0            0                  0               2   
3               0            1                 44              27   
4               0            0                  2              31   
...           ...          ...                ...             ...   
16056           0            0                  0               4   
16057           0            1                 22               9   
16058           0            0                  0              27   
16059           0            0                  0              21   
16060           0            0                  0               0   

       number_of_table  url_length  has_hyphens  
0                    1           0            0  
1                    0           0            0  
2                    0           0            0  
3                    0           0            0  
4                    0           0            0  
...                ...         ...          ...  
16056                2           0            0  
16057                0           0            0  
16058                0           0            0  
16059                0           0            0  
16060                0           0            0  

[1

In [56]:
legit["has_hyphens"].value_counts()

has_hyphens
0    15466
1      595
Name: count, dtype: int64

In [57]:
phish["has_hyphens"] = phish["URL"].apply(has_hyphens)

In [58]:
phish

URL  has_title  \
0      http://www.eki-net.con-aescceccesaas.qfuhtb.to...          0   
1      http://www.eki-net.con-aescceccesaas.qsnbpy.to...          0   
2      http://www.eki-net.con-aescceccesaas.trea.top/...          0   
3      http://www.eki-net.con-aescceccesaas.ucwxyw.to...          0   
4      http://www.eki-net.con-aescceccesaas.vcuu.top/...          0   
...                                                  ...        ...   
10519                              https://dloodles.app/          1   
10520                             https://artblooks.xyz/          1   
10521                           http://ohnizwl.htswyx.cn          1   
10522                https://poocoinect-tk.blogspot.com/          1   
10523                    https://osmosis.sezenradyo.com/          1   

       has_input  has_button  has_image  has_submit  has_link  has_password  \
0              0           0          0           0         0             0   
1              0           0          0           0         0             0   
2              0           0          0           0         0             0   
3              0           0          0           0         0             0   
4              0           0          0           0         0             0   
...          ...         ...        ...         ...       ...           ...   
10519          0           0          0           0         0             0   
10520          0           0          0           0         1             0   
10521          0           0          0           0         0             0   
10522          0           1          0           0         1             0   
10523          0           1          0           0         1             0   

       has_email_input  has_hidden_element  ...  has_text_input  \
0                    0                   0  ...               0   
1                    0                   0  ...               0   
2                    0                   0  ...               0   
3                    0                   0  ...               0   
4                    0                   0  ...               0   
...                ...                 ...  ...             ...   
10519                0                   0  ...               0   
10520                0                   0  ...               0   
10521                0                   0  ...               0   
10522                0                   0  ...               0   
10523                0                   0  ...               0   

       number_of_meta  has_nav  has_object  has_picture  number_of_sources  \
0                   0        0           0            0                  0   
1                   0        0           0            0                  0   
2                   0        0           0            0                  0   
3                   0        0           0            0                  0   
4                   0        0           0            0                  0   
...               ...      ...         ...          ...                ...   
10519               0        0           0            0                  0   
10520              11        0           0            0                  0   
10521               1        0           0            0                  0   
10522               4        1           0            0                  0   
10523               3        1           0            0                  0   

       number_of_span  number_of_table  url_length  has_hyphens  
0                   0                0           1            1  
1                   0                0           1            1  
2                   0                0           0            1  
3                   0                0           1            1  
4                   0                0           0            1  
...               ...              ...         ...          ...  
10519               0                1         

In [59]:
phish['has_hyphens'].value_counts()

has_hyphens
0    5794
1    4730
Name: count, dtype: int64

<h5>Data story</h5> 
   
   * In the phishing URLs, there are 5794 rows without '-' and 4730 rows with '-' hyphens.
   * In the legitimate URLs, there are 15466 rows without hyphens '-' and 595 rows with hyphens '-'
   * Therefore, hyphen checks will be a crucial feature for training the machine learning model in this project

<h3> IP Address in the URL</h3>

*  Checks for the presence of IP address in the URL. URLs may have IP address instead of domain     name. If an IP address is used as an alternative of the domain name in the URL, we can be sure    that someone is trying to steal personal information with this URL.
*  If the domain part of URL has IP address, the value assigned to this feature is 1 (phishing)      or else 0 (legitimate).

In [60]:
import ipaddress # imported ip address module for getting the ip out of the url

def having_ip(url):
    try:
        parsed_url = urlparse(url)
        ipaddress.ip_address(parsed_url.netloc)#using ipaddress.ip_address we will be checking whether the url have the ip address in it or not
        return 1
    except ValueError:
        return 0


In [61]:
phish["having_ip"] = phish["URL"].apply(having_ip)

In [62]:
phish["having_ip"].value_counts()

having_ip
0    10520
1        4
Name: count, dtype: int64

In [63]:
legit["having_ip"] = legit["URL"].apply(having_ip)

In [64]:
legit['having_ip'].value_counts()

having_ip
0    16061
Name: count, dtype: int64

In [65]:
legit.drop('having_ip',inplace=True,axis=1)

In [66]:
phish.drop('having_ip',inplace=True,axis=1)

<h5>Data story</h5>

*  Phishing URLs: Out of the total phishing URLs, only 4 URLs contain an IP address, while the       remaining 10,520 URLs do not have any IP address in the URL
*  Legitimate URLs: None of the legitimate URLs contain an IP address in the URL.
*  Conclusion: Given the minimal presence of IP addresses and their lack of defining significance, it's concluded that this column does not provide         meaningful insights. Therefore, it will be dropped from further analysis

<h3> "@" Symbol in URL</h3>

*   Checks for the presence of '@' symbol in the URL. Using “@” symbol in the URL leads the browser to ignore everything preceding the “@” symbol and      the real address often follows the “@” symbol.
*   If the URL has '@' symbol, the value assigned to this feature is 1 (phishing) or else 0 (legitimate).

In [67]:
def having_atsign(url):
    if '@' in urlparse(url).netloc:
        return 1
    else:
        return 0

In [68]:
phish["atsign"] = phish['URL'].apply(having_atsign)

In [69]:
phish["atsign"].value_counts()

atsign
0    10524
Name: count, dtype: int64

In [70]:
legit["atsign"] = legit['URL'].apply(having_atsign)

In [71]:
legit['atsign'].value_counts()

atsign
0    16061
Name: count, dtype: int64

In [72]:
legit.drop("atsign",inplace=True,axis=1)

In [73]:
phish.drop("atsign",inplace=True,axis=1)

<h5>Data story</h5>

*  In both phishing and the legitimate url's there is no '@' in both
*  it's concluded that this column does not provide meaningful insights. Therefore, it will be dropped from further analysis

<h3>Extracting domain out of url and finding the length of the domain</h3>

In [74]:
def extract_domain(url):
    parsed_url = urlparse(url)
    return parsed_url.netloc

In [75]:
legit['domain'] = legit["URL"].apply(extract_domain)

In [ ]:
legit['domain_length'] = legit["domain"].apply(len)

In [76]:
legit

URL  has_title  has_input  has_button  has_image  \
0            http://google.com          1          1           0          0   
1           http://youtube.com          1          1           0          0   
2          http://facebook.com          1          1           1          0   
3         http://microsoft.com          1          1           1          0   
4           http://netflix.com          1          1           1          0   
...                        ...        ...        ...         ...        ...   
16056       http://emis.gov.eg          1          1           0          0   
16057     http://hilton.com.tr          1          1           1          0   
16058  http://medicines.org.uk          1          1           1          0   
16059        http://piwigo.org          1          0           1          0   
16060        http://koddos.com          1          0           0          0   

       has_submit  has_link  has_password  has_email_input  \
0               1         0             0                0   
1               0         1             0                0   
2               0         1             1                0   
3               0         1             0                0   
4               0         1             0                1   
...           ...       ...           ...              ...   
16056           0         1             0                0   
16057           0         1             0                0   
16058           0         1             0                0   
16059           0         1             0                0   
16060           0         0             0                0   

       has_hidden_element  ...  number_of_meta  has_nav  has_object  \
0                       1  ...               2        0           0   
1                       0  ...               7        0           0   
2                       1  ...              18        0           0   
3                       0  ...              26        1           0   
4                       0  ...              20        0           0   
...                   ...  ...             ...      ...         ...   
16056                   1  ...               1        1           0   
16057                   1  ...               5        1           0   
16058                   0  ...               6        1           0   
16059                   0  ...               4        1           0   
16060                   0  ...               0        0           0   

       has_picture  number_of_sources  number_of_span  number_of_table  \
0                0                  0               8                1   
1                0                  0               1                0   
2                0                  0               2                0   
3                1                 44              27                0   
4                0                  2              31                0   
...            ...                ...             ...              ...   
16056            0                  0               4                2   
16057            1                 22               9                0   
16058            0                  0              27                0   
16059            0                  0              21                0   
16060            0                  0               0                0   

       url_length  has_hyphens            domain  
0               0            0        google.com  
1               0            0       youtube.com  
2               0            0      facebook.com  
3               0            0     microsoft.com  
4               0            0       netflix.com  
...           ...          ...               ...  
16056           0            0       emis.gov.eg  
16057           0            0     hilton.com.tr  
16058           0            0  medicines.org.uk  
16059           0            0        piwigo.org  
16060           0            0

In [77]:
phish['domain'] = phish["URL"].apply(extract_domain)

In [78]:
phish

URL  has_title  \
0      http://www.eki-net.con-aescceccesaas.qfuhtb.to...          0   
1      http://www.eki-net.con-aescceccesaas.qsnbpy.to...          0   
2      http://www.eki-net.con-aescceccesaas.trea.top/...          0   
3      http://www.eki-net.con-aescceccesaas.ucwxyw.to...          0   
4      http://www.eki-net.con-aescceccesaas.vcuu.top/...          0   
...                                                  ...        ...   
10519                              https://dloodles.app/          1   
10520                             https://artblooks.xyz/          1   
10521                           http://ohnizwl.htswyx.cn          1   
10522                https://poocoinect-tk.blogspot.com/          1   
10523                    https://osmosis.sezenradyo.com/          1   

       has_input  has_button  has_image  has_submit  has_link  has_password  \
0              0           0          0           0         0             0   
1              0           0          0           0         0             0   
2              0           0          0           0         0             0   
3              0           0          0           0         0             0   
4              0           0          0           0         0             0   
...          ...         ...        ...         ...       ...           ...   
10519          0           0          0           0         0             0   
10520          0           0          0           0         1             0   
10521          0           0          0           0         0             0   
10522          0           1          0           0         1             0   
10523          0           1          0           0         1             0   

       has_email_input  has_hidden_element  ...  number_of_meta  has_nav  \
0                    0                   0  ...               0        0   
1                    0                   0  ...               0        0   
2                    0                   0  ...               0        0   
3                    0                   0  ...               0        0   
4                    0                   0  ...               0        0   
...                ...                 ...  ...             ...      ...   
10519                0                   0  ...               0        0   
10520                0                   0  ...              11        0   
10521                0                   0  ...               1        0   
10522                0                   0  ...               4        1   
10523                0                   0  ...               3        1   

       has_object  has_picture  number_of_sources  number_of_span  \
0               0            0                  0               0   
1               0            0                  0               0   
2               0            0                  0               0   
3               0            0                  0               0   
4               0            0                  0               0   
...           ...          ...                ...             ...   
10519           0            0                  0               0   
10520           0            0                  0              14   
10521           0            0                  0               0   
10522           0            0                  0              51   
10523           0            0                  0               0   

       number_of_table  url_length  has_hyphens  \
0                    0           1            1   
1                    0           1            1   
2                    0           0            1   
3                    0           1            1   
4                    0           0            1   
...                ...         ...          ...   
10519                1           0            0   
10520                0           0            0   
10521                0           0            0   

<h3>Subdomain Count</h3>


In [79]:
import tldextract
def count_subdomains(url):
    # Use tldextract library to extract subdomains
    ext = tldextract.extract(url)
    subdomains = ext.subdomain.split('.')  # Split subdomains by dot separator
    # Return the count of subdomains
    return len(subdomains)

In [80]:
phish['subdomain_count'] = phish['URL'].apply(count_subdomains)

unable to cache publicsuffix.org-tlds.{'urls': ('https://publicsuffix.org/list/public_suffix_list.dat', 'https://raw.githubusercontent.com/publicsuffix/list/master/public_suffix_list.dat'), 'fallback_to_snapshot': True} in C:\Program Files\anaconda\Lib\site-packages\tldextract\.suffix_cache/publicsuffix.org-tlds\de84b5ca2167d4c83e38fb162f2e8738.tldextract.json. This could refresh the Public Suffix List over HTTP every app startup. Construct your `TLDExtract` with a writable `cache_dir` or set `cache_dir=False` to silence this warning. [WinError 5] Access is denied: 'C:\\Program Files\\anaconda\\Lib\\site-packages\\tldextract\\.suffix_cache'


In [81]:
phish

URL  has_title  \
0      http://www.eki-net.con-aescceccesaas.qfuhtb.to...          0   
1      http://www.eki-net.con-aescceccesaas.qsnbpy.to...          0   
2      http://www.eki-net.con-aescceccesaas.trea.top/...          0   
3      http://www.eki-net.con-aescceccesaas.ucwxyw.to...          0   
4      http://www.eki-net.con-aescceccesaas.vcuu.top/...          0   
...                                                  ...        ...   
10519                              https://dloodles.app/          1   
10520                             https://artblooks.xyz/          1   
10521                           http://ohnizwl.htswyx.cn          1   
10522                https://poocoinect-tk.blogspot.com/          1   
10523                    https://osmosis.sezenradyo.com/          1   

       has_input  has_button  has_image  has_submit  has_link  has_password  \
0              0           0          0           0         0             0   
1              0           0          0           0         0             0   
2              0           0          0           0         0             0   
3              0           0          0           0         0             0   
4              0           0          0           0         0             0   
...          ...         ...        ...         ...       ...           ...   
10519          0           0          0           0         0             0   
10520          0           0          0           0         1             0   
10521          0           0          0           0         0             0   
10522          0           1          0           0         1             0   
10523          0           1          0           0         1             0   

       has_email_input  has_hidden_element  ...  has_nav  has_object  \
0                    0                   0  ...        0           0   
1                    0                   0  ...        0           0   
2                    0                   0  ...        0           0   
3                    0                   0  ...        0           0   
4                    0                   0  ...        0           0   
...                ...                 ...  ...      ...         ...   
10519                0                   0  ...        0           0   
10520                0                   0  ...        0           0   
10521                0                   0  ...        0           0   
10522                0                   0  ...        1           0   
10523                0                   0  ...        1           0   

       has_picture  number_of_sources  number_of_span  number_of_table  \
0                0                  0               0                0   
1                0                  0               0                0   
2                0                  0               0                0   
3                0                  0               0                0   
4                0                  0               0                0   
...            ...                ...             ...              ...   
10519            0                  0               0                1   
10520            0                  0              14                0   
10521            0                  0               0                0   
10522            0                  0              51                0   
10523            0                  0               0                0   

       url_length  has_hyphens                                    domain  \
0               1            1  www.eki-net.con-aescceccesaas.qfuhtb.top   
1               1            1  www.eki-net.con-aescceccesaas.qsnbpy.top   
2               0            1    www.eki-net.con-aescceccesaas.trea.top   
3               1            1  www.eki-net.con-aescceccesaas.ucwxyw.top   
4               0            1    www.eki-net.con-aescceccesaas.vcuu.top   
...           ...          ...           

In [82]:
phish['subdomain_count'].value_counts()

subdomain_count
3    7471
2    2450
1     582
5      17
6       3
4       1
Name: count, dtype: int64

In [83]:
legit['subdomain_count'] = legit['URL'].apply(count_subdomains)

In [84]:
legit['subdomain_count'].value_counts()

subdomain_count
1    16061
Name: count, dtype: int64

In [85]:
legit

URL  has_title  has_input  has_button  has_image  \
0            http://google.com          1          1           0          0   
1           http://youtube.com          1          1           0          0   
2          http://facebook.com          1          1           1          0   
3         http://microsoft.com          1          1           1          0   
4           http://netflix.com          1          1           1          0   
...                        ...        ...        ...         ...        ...   
16056       http://emis.gov.eg          1          1           0          0   
16057     http://hilton.com.tr          1          1           1          0   
16058  http://medicines.org.uk          1          1           1          0   
16059        http://piwigo.org          1          0           1          0   
16060        http://koddos.com          1          0           0          0   

       has_submit  has_link  has_password  has_email_input  \
0               1         0             0                0   
1               0         1             0                0   
2               0         1             1                0   
3               0         1             0                0   
4               0         1             0                1   
...           ...       ...           ...              ...   
16056           0         1             0                0   
16057           0         1             0                0   
16058           0         1             0                0   
16059           0         1             0                0   
16060           0         0             0                0   

       has_hidden_element  ...  has_nav  has_object  has_picture  \
0                       1  ...        0           0            0   
1                       0  ...        0           0            0   
2                       1  ...        0           0            0   
3                       0  ...        1           0            1   
4                       0  ...        0           0            0   
...                   ...  ...      ...         ...          ...   
16056                   1  ...        1           0            0   
16057                   1  ...        1           0            1   
16058                   0  ...        1           0            0   
16059                   0  ...        1           0            0   
16060                   0  ...        0           0            0   

       number_of_sources  number_of_span  number_of_table  url_length  \
0                      0               8                1           0   
1                      0               1                0           0   
2                      0               2                0           0   
3                     44              27                0           0   
4                      2              31                0           0   
...                  ...             ...              ...         ...   
16056                  0               4                2           0   
16057                 22               9                0           0   
16058                  0              27                0           0   
16059                  0              21                0           0   
16060                  0               0                0           0   

       has_hyphens            domain  subdomain_count  
0                0        google.com                1  
1                0       youtube.com                1  
2                0      facebook.com                1  
3                0     microsoft.com                1  
4                0       netflix.com                1  
...            ...               ...              ...  
16056            0       emis.gov.eg                1  
16057            0     hilton.com.tr                1  
16058            0  medicines.org.uk                1  
16059            0        piwigo.org                1  
16060            0     

<h5>Data story</h5>

*  Phishing URLs:
      *   7471 URLs have 3 subdomains.
      *   2450 URLs have 2 subdomains.
      *   582 URLs have 1 subdomain.
      *   Only a few URLs have 5, 6, and 4 subdomains, with 17, 3, and 1 URL(s) respectively.

*  Legitimate URLs
      *  The majority of legitimate URLs (16061) have only 1 subdomain.

*  Conclusion:
   *  
In summary, the number of subdomains in a URL can help identify whether it's a phishing or legitimate website. Phishing sites often have many subdomains, possibly to hide their true identity, while legitimate sites usually have fewer. This information could be really useful for creating better tools to spot phishing websites.     
    .



.






nely.


<h3>
Top-Level Domain (TLD) Type</h3

* Identify the type of TLD used (e.g.,top,info .org, .xyz, etc.). Certain TLDs are more commonly associated with phishing.>

In [86]:
def mark_phishing_tld(url):
    ext = tldextract.extract(url)
    phishing_tlds = ['xyz', 'top', 'info', 'loan', 'xmr', 'ink', 'sk', 'ag', 'ryukyu', 'ltc', 'gr', 'fund', 'sx']

    # Convert both TLD and phishing_tlds list to lowercase for case-insensitive comparison
    tld = ext.suffix.lower()
    phishing_tlds = [tld.lower() for tld in phishing_tlds]

    if tld in phishing_tlds:
        return 1  # Mark as phishing
    else:
        return 0  # Mark as legitimate

In [87]:
phish['Phishing_TLD'] = phish['URL'].apply(mark_phishing_tld)

In [88]:
phish['Phishing_TLD'].value_counts()

Phishing_TLD
0    5907
1    4617
Name: count, dtype: int64

In [89]:
legit['Phishing_TLD'] = legit['URL'].apply(mark_phishing_tld)

In [90]:
legit['Phishing_TLD'].value_counts()

Phishing_TLD
0    15942
1      119
Name: count, dtype: int64

In [91]:
phish

URL  has_title  \
0      http://www.eki-net.con-aescceccesaas.qfuhtb.to...          0   
1      http://www.eki-net.con-aescceccesaas.qsnbpy.to...          0   
2      http://www.eki-net.con-aescceccesaas.trea.top/...          0   
3      http://www.eki-net.con-aescceccesaas.ucwxyw.to...          0   
4      http://www.eki-net.con-aescceccesaas.vcuu.top/...          0   
...                                                  ...        ...   
10519                              https://dloodles.app/          1   
10520                             https://artblooks.xyz/          1   
10521                           http://ohnizwl.htswyx.cn          1   
10522                https://poocoinect-tk.blogspot.com/          1   
10523                    https://osmosis.sezenradyo.com/          1   

       has_input  has_button  has_image  has_submit  has_link  has_password  \
0              0           0          0           0         0             0   
1              0           0          0           0         0             0   
2              0           0          0           0         0             0   
3              0           0          0           0         0             0   
4              0           0          0           0         0             0   
...          ...         ...        ...         ...       ...           ...   
10519          0           0          0           0         0             0   
10520          0           0          0           0         1             0   
10521          0           0          0           0         0             0   
10522          0           1          0           0         1             0   
10523          0           1          0           0         1             0   

       has_email_input  has_hidden_element  ...  has_object  has_picture  \
0                    0                   0  ...           0            0   
1                    0                   0  ...           0            0   
2                    0                   0  ...           0            0   
3                    0                   0  ...           0            0   
4                    0                   0  ...           0            0   
...                ...                 ...  ...         ...          ...   
10519                0                   0  ...           0            0   
10520                0                   0  ...           0            0   
10521                0                   0  ...           0            0   
10522                0                   0  ...           0            0   
10523                0                   0  ...           0            0   

       number_of_sources  number_of_span  number_of_table  url_length  \
0                      0               0                0           1   
1                      0               0                0           1   
2                      0               0                0           0   
3                      0               0                0           1   
4                      0               0                0           0   
...                  ...             ...              ...         ...   
10519                  0               0                1           0   
10520                  0              14                0           0   
10521                  0               0                0           0   
10522                  0              51                0           0   
10523                  0               0                0           0   

       has_hyphens                                    domain  subdomain_count  \
0                1  www.eki-net.con-aescceccesaas.qfuhtb.top                3   
1                1  www.eki-net.con-aescceccesaas.qsnbpy.top                3   
2                1    www.eki-net.con-aescceccesaas.trea.top                3   
3                1  www.eki-net.con-aescceccesaas.ucwxyw.top                3   
4                1    www.eki-net.con-aescceccesaas.vcuu

<h5>Data story</h5>

<b>Phishing Dataset Analysis</b>
*  
In the 'phish' dataset, which contains URLs suspected of being phishing, the 'Phishing_TLD' column shows that 4,617 URL%) are classified a     s phishing (1) based on their TLD, while 5,907 URL%) are classified as legitimate (0

<b>Legitimate Dataset Analysis</b>
*  
In the 'legit' dataset, which contains URLs believed to be legitimate, the 'Phishing_TLD' column shows that 15,942 URL%) are classified a      s legitimate (0) based on their TLD, while 119 URL%) are classified as phishing (1

  <b>Insights and Limitations</b>
*  
The presence of 119 URLs classified as phishing (1) in the 'legit' dataset could be due to false positives, limitations of the TLD list, o            r potential mislabeling in the dataset.).).

In [92]:
phish

URL  has_title  \
0      http://www.eki-net.con-aescceccesaas.qfuhtb.to...          0   
1      http://www.eki-net.con-aescceccesaas.qsnbpy.to...          0   
2      http://www.eki-net.con-aescceccesaas.trea.top/...          0   
3      http://www.eki-net.con-aescceccesaas.ucwxyw.to...          0   
4      http://www.eki-net.con-aescceccesaas.vcuu.top/...          0   
...                                                  ...        ...   
10519                              https://dloodles.app/          1   
10520                             https://artblooks.xyz/          1   
10521                           http://ohnizwl.htswyx.cn          1   
10522                https://poocoinect-tk.blogspot.com/          1   
10523                    https://osmosis.sezenradyo.com/          1   

       has_input  has_button  has_image  has_submit  has_link  has_password  \
0              0           0          0           0         0             0   
1              0           0          0           0         0             0   
2              0           0          0           0         0             0   
3              0           0          0           0         0             0   
4              0           0          0           0         0             0   
...          ...         ...        ...         ...       ...           ...   
10519          0           0          0           0         0             0   
10520          0           0          0           0         1             0   
10521          0           0          0           0         0             0   
10522          0           1          0           0         1             0   
10523          0           1          0           0         1             0   

       has_email_input  has_hidden_element  ...  has_object  has_picture  \
0                    0                   0  ...           0            0   
1                    0                   0  ...           0            0   
2                    0                   0  ...           0            0   
3                    0                   0  ...           0            0   
4                    0                   0  ...           0            0   
...                ...                 ...  ...         ...          ...   
10519                0                   0  ...           0            0   
10520                0                   0  ...           0            0   
10521                0                   0  ...           0            0   
10522                0                   0  ...           0            0   
10523                0                   0  ...           0            0   

       number_of_sources  number_of_span  number_of_table  url_length  \
0                      0               0                0           1   
1                      0               0                0           1   
2                      0               0                0           0   
3                      0               0                0           1   
4                      0               0                0           0   
...                  ...             ...              ...         ...   
10519                  0               0                1           0   
10520                  0              14                0           0   
10521                  0               0                0           0   
10522                  0              51                0           0   
10523                  0               0                0           0   

       has_hyphens                                    domain  subdomain_count  \
0                1  www.eki-net.con-aescceccesaas.qfuhtb.top                3   
1                1  www.eki-net.con-aescceccesaas.qsnbpy.top                3   
2                1    www.eki-net.con-aescceccesaas.trea.top                3   
3                1  www.eki-net.con-aescceccesaas.ucwxyw.top                3   
4                1    www.eki-net.con-aescceccesaas.vcuu

In [93]:
phish['label'] = 1  # Add label column for phishing dataset



In [94]:
legit['label'] = 0  # Add label column for legitimate dataset

In [99]:
phish

URL  has_title  \
0      http://www.eki-net.con-aescceccesaas.qfuhtb.to...          0   
1      http://www.eki-net.con-aescceccesaas.qsnbpy.to...          0   
2      http://www.eki-net.con-aescceccesaas.trea.top/...          0   
3      http://www.eki-net.con-aescceccesaas.ucwxyw.to...          0   
4      http://www.eki-net.con-aescceccesaas.vcuu.top/...          0   
...                                                  ...        ...   
10519                              https://dloodles.app/          1   
10520                             https://artblooks.xyz/          1   
10521                           http://ohnizwl.htswyx.cn          1   
10522                https://poocoinect-tk.blogspot.com/          1   
10523                    https://osmosis.sezenradyo.com/          1   

       has_input  has_button  has_image  has_submit  has_link  has_password  \
0              0           0          0           0         0             0   
1              0           0          0           0         0             0   
2              0           0          0           0         0             0   
3              0           0          0           0         0             0   
4              0           0          0           0         0             0   
...          ...         ...        ...         ...       ...           ...   
10519          0           0          0           0         0             0   
10520          0           0          0           0         1             0   
10521          0           0          0           0         0             0   
10522          0           1          0           0         1             0   
10523          0           1          0           0         1             0   

       has_email_input  has_hidden_element  ...  has_picture  \
0                    0                   0  ...            0   
1                    0                   0  ...            0   
2                    0                   0  ...            0   
3                    0                   0  ...            0   
4                    0                   0  ...            0   
...                ...                 ...  ...          ...   
10519                0                   0  ...            0   
10520                0                   0  ...            0   
10521                0                   0  ...            0   
10522                0                   0  ...            0   
10523                0                   0  ...            0   

       number_of_sources  number_of_span  number_of_table  url_length  \
0                      0               0                0           1   
1                      0               0                0           1   
2                      0               0                0           0   
3                      0               0                0           1   
4                      0               0                0           0   
...                  ...             ...              ...         ...   
10519                  0               0                1           0   
10520                  0              14                0           0   
10521                  0               0                0           0   
10522                  0              51                0           0   
10523                  0               0                0           0   

       has_hyphens                                    domain  subdomain_count  \
0                1  www.eki-net.con-aescceccesaas.qfuhtb.top                3   
1                1  www.eki-net.con-aescceccesaas.qsnbpy.top                3   
2                1    www.eki-net.con-aescceccesaas.trea.top                3   
3                1  www.eki-net.con-aescceccesaas.ucwxyw.top                3   
4                1    www.eki-net.con-aescceccesaas.vcuu.top                3   
...            ...                                       ...              ...   
10519            0                    

In [100]:
legit.shape

(16061, 50)

In [101]:
phish.shape

(10524, 50)

In [102]:
"""combined_df = pd.concat([legit,phish],ignore_index=True)"""

In [103]:
"""shuffled_df = combined_df.sample(frac=1, random_state=42).reset_index(drop=True)"""

In [104]:
"""shuffled_df.to_csv('phishing_detection_combined_data.csv', index=False)"""

In [105]:
"""df =pd.read_csv("phishing_detection_combined_data.csv")"""

In [106]:
"""df"""

URL  has_title  \
0                                       http://dng65.com          1   
1                                   http://mediaroom.com          1   
2                                   http://yieldlift.com          1   
3                                    http://yowindow.com          1   
4                              http://brighthorizons.com          1   
...                                                  ...        ...   
26580  http://www.vivcsiiaaaicca.vicsvesai.lzvsi5.icu...          0   
26581                             http://lightwidget.com          1   
26582                                    http://agora.io          1   
26583                                     http://lnk.bio          1   
26584         http://www.macesareod.bwlrdn.icu/page1.php          0   

       has_input  has_button  has_image  has_submit  has_link  has_password  \
0              1           0          0           1         1             0   
1              1           1          0           0         1             0   
2              1           1          0           0         1             0   
3              0           0          0           0         1             0   
4              1           1          0           0         1             0   
...          ...         ...        ...         ...       ...           ...   
26580          0           0          0           0         0             0   
26581          1           1          0           0         1             0   
26582          0           1          0           0         1             0   
26583          0           1          0           0         1             0   
26584          0           0          0           0         0             0   

       has_email_input  has_hidden_element  ...  has_picture  \
0                    0                   0  ...            0   
1                    0                   1  ...            0   
2                    1                   1  ...            0   
3                    0                   0  ...            0   
4                    0                   1  ...            0   
...                ...                 ...  ...          ...   
26580                0                   0  ...            0   
26581                0                   0  ...            1   
26582                0                   0  ...            0   
26583                0                   0  ...            1   
26584                0                   0  ...            0   

       number_of_sources  number_of_span  number_of_table  url_length  \
0                      0               1                0           0   
1                      2              24                0           0   
2                      0               8                0           0   
3                      0               8                0           0   
4                      0              38                0           0   
...                  ...             ...              ...         ...   
26580                  0               0                0           1   
26581                 10               0                0           0   
26582                  0              19                0           0   
26583                 10              20                0           0   
26584                  0               0                0           0   

       has_hyphens                                   domain  subdomain_count  \
0                0                                dng65.com                1   
1                0                            mediaroom.com                1   
2                0                            yieldlift.com                1   
3                0                             yowindow.com                1   
4                0                       brighthorizons.com                1   
...            ...                                      ...              ...   
26580            0  www.vivcsiiaaaicca.vicsve